remix-analyzer/src/solidity-analyzer/modules/gasCosts.ts

```diff
@@ -1,72 +1,96 @@
-        if (functionName === '') {
-          continue
-        }
-        const gas: string = contract.object.evm.gasEstimates.external[functionName]
-        const gasString: string = gas === null ? 'unknown or not constant' : 'high: ' + gas
-        if (gas === null || parseInt(gas) >= 3000000 || gas === 'infinite') {
-          report.push({
-            warning: `Gas requirement of function ${contract.name}.${functionName} ${gasString}. 
-            If the gas requirement of a function is higher than the block gas limit, it cannot be executed.
-            Please avoid loops in your functions or actions that modify large areas of storage
-            (this includes clearing or copying arrays in storage)`
-          })
+    })
+    for (const method of methodsWithSignature) {
+      for (const filename in compilationResults.contracts) {
+        for (const contractName in compilationResults.contracts[filename]) {
+          const contract: CompiledContract = compilationResults.contracts[filename][contractName]
+          const methodGas: Record<string, any> | undefined = this.checkMethodGas(contract, method.signature)
+          if(methodGas && methodGas.isInfinite) {
+            if(methodGas.isFallback) {
+              report.push({
+                warning: `Fallback function of contract ${contractName} requires too much gas (${methodGas.msg}). 
+                If the fallback function requires more than 2300 gas, the contract cannot receive Ether.`,
+                location: method.src
+              })
+            } else {
+              report.push({
+                warning: `Gas requirement of function ${contractName}.${method.name} ${methodGas.msg}. 
+                If the gas requirement of a function is higher than the block gas limit, it cannot be executed.
+                Please avoid loops in your functions or actions that modify large areas of storage
+                (this includes clearing or copying arrays in storage)`,
+                location: method.src
+              })
+            } 
+          } else continue
         }
       }
-    })
+    }
     return report
   }
 
-  /**
-  * call the given @arg cb (function) for all the contracts. Uses last compilation result
-  * stop visiting when cb return true
-  * @param {Function} cb    - callback
-  */
-  // @TODO has been copied from remix-ide repo ! should fix that soon !
-  private visitContracts (contracts: CompiledContractObj | undefined, cb: ((contract: VisitedContract) => void | undefined)): void {
-    for (let file in contracts) {
-      for (let name in contracts[file]) {
-        if (cb({ name: name, object: contracts[file][name], file: file })) return
+  // To create the method signature similar to contract.evm.gasEstimates.external object
+  // For address payable, return address 
+  private getSplittedTypeDesc(node: VariableDeclarationAstNode): string {
+    return node.typeDescriptions.typeString.split(' ')[0]
```

Comment:
Can we have unit test for that which include complex type (like struct and array).

e.g `function  testWithArray (bytes32[] memory param) public {}` is translated to `"testWithArray(bytes32[])": "infinite",`

and

`function  testWithArrayOfStruct (Proposal[] memory param) public {}` is translated to `"testWithArrayOfStruct((bytes32,uint256)[])": "infinite",` (`(bytes32,uint256)` being a `Proposal`)

Reply:
ok, will do it.

Reply:
This is done.